openpype/hosts/houdini/api/lib.py

```diff
@@ -747,3 +753,60 @@ def get_camera_from_container(container):
 
     assert len(cameras) == 1, "Camera instance must have only one camera"
     return cameras[0]
+
+
+def update_houdini_vars_context():
+    """Update Houdini vars to match current context.
+
+    This will only do something if the setting is enabled in project settings.
+    """
+
+    project_settings = get_current_project_settings()
+    houdini_vars_settings = \
+        project_settings["houdini"]["general"]["update_houdini_var_context"]
+
+    if houdini_vars_settings["enabled"]:
+        houdini_vars = houdini_vars_settings["houdini_vars"]
+
+        # Remap AYON settings structure to OpenPype settings structure
+        # It allows me to use the same logic for both AYON and OpenPype
+        if isinstance(houdini_vars, list):
+            items = {}
+            for item in houdini_vars:
+                items.update({item["var"]: item["path"]})
+
+            houdini_vars = items
+
+        for var, path in houdini_vars.items():
+            # get and resolve job path template
+            path = StringTemplate.format_template(
+                path,
+                get_current_context_template_data()
+            )
+            path = path.replace("\\", "/")
+
+            if var == "JOB" and path == "":
+                # sync $JOB to $HIP if $JOB is empty
+                path = os.environ["HIP"]
+
+            current_path = hou.hscript("echo -n `${}`".format(var))[0]
+
+            # sync both environment variables.
+            # because houdini doesn't do that by default
+            # on opening new files
+            os.environ[var] = current_path
+
+            if current_path != path:
+                hou.hscript("set {}={}".format(var, path))
+                os.environ[var] = path
+
+                try:
+                    os.makedirs(path)
+                except OSError as e:
+                    if e.errno != errno.EEXIST:
+                        print(
+                            "  - Failed to create {} dir. Maybe due to "
+                            "insufficient permissions.".format(var)
+                        )
+
+                print("  - Updated ${} to {}".format(var, path))
```

Comment:
Since as per my examples provided when explaining why we could benefit from expanding this functionality for user requests to set vars for asset resolution, frame ranges, etc. the vars set through this function now aren't always actual paths on disk - but could be other arbitrary data, like a frame start value. So likely we need an extra toggle per var in settings that says "Treat as path and create folder when it does not exist" so that you can enable it for vars that are supposed to be absolute paths.

And likely we'd need to refactor the function to something like this:
```suggestion


def update_houdini_vars_context():
    """Update Houdini vars to match current context based.

    This will only do something if the setting is enabled in project settings.

    """

    project_settings = get_current_project_settings()
    houdini_vars_settings = \
        project_settings["houdini"]["general"]["update_houdini_var_context"]

    if not houdini_vars_settings["enabled"]:
        return

    houdini_vars = houdini_vars_settings["houdini_vars"]

    # Remap AYON settings structure to OpenPype settings structure
    # It allows me to use the same logic for both AYON and OpenPype
    if isinstance(houdini_vars, list):
        items = {}
        for item in houdini_vars:
            items.update({item["var"]: item["path"]})

        houdini_vars = items

    if not houdini_vars:
        # No vars specified - nothing to do
        return

    # For consistency reasons we always force all vars to be uppercase
    houdini_vars = {key.upper(): value for key, value in houdini_vars.items()}

    # TODO: I've now NOT used the separate function you made for this since
    #   now we're accessing the asset document itself as well.
    # Retrieve any context data for formatting
    # pre-prepare get_template_data args
    current_context = get_current_context()
    project_name = current_context["project_name"]
    asset_name = current_context["asset_name"]
    anatomy = Anatomy(project_name)

    # prepare get_template_data args
    project_doc = get_project(project_name)
    asset_doc = get_asset_by_name(project_name, asset_name)
    task_name = current_context["task_name"]
    host_name = get_current_host_name()

    # get template data
    template_data = get_template_data(
        project_doc, asset_doc, task_name, host_name
    )
    template_data["root"] = anatomy.roots

    # TODO: We add to documentation asset data examples that an admin might
    #    want to use.
    # To allow setting vars to context specific frame ranges, resolution
    # we need to provide more than the regular template data
    # By doing this you can access data like:
    # - {assetData[frameStart]}
    # - {assetData[frameEnd]}
    # - {assetData[handleStart]}
    # - {assetData[handleEnd]}
    # - {assetData[frameStartHandle]}
    # - {assetData[frameEndHandle]}
    # - {assetData[resolutionHeight]}
    # - {assetData[resolutionWidth]}
    # This would also expose e.g. (if ftrack module is enabled/synced):
    # - {assetData[ftrackId]}
    asset_data = asset_doc["data"].copy()
    template_data["asset_data"] = asset_data
    # The `frameStartHandle` and `frameEndHandle` doesn't exist in the
    # document's data thus we always need to compute it so it can be used
    # as formattable data
    if "frameStart" in asset_data and "handleStart" in asset_data:
        asset_data["frameStartHandle"] = (
                asset_data["frameStart"] - asset_data["frameStartHandle"]
        )
    if "frameEnd" in asset_data and "handleEnd" in asset_data:
        asset_data["frameEndHandle"] = (
                asset_data["frameEnd"] + asset_data["handleEnd"]
        )
    template_data["assetData"] = asset_data

    for var, template_value in houdini_vars.items():
        # get template and format with template data
        value = StringTemplate.format_template(template_value, template_data)

        if var == "JOB" and path == "":
            # sync $JOB to $HIP if $JOB is empty
            path = os.environ["HIP"]

        # TODO: Allow each entry in settings to specify whether it should
        #       be treated as a path (and with that also force the folder
        #       creation.
        if var_is_path:
            value = value.replace("\\", "/")

            # Force create folder when treating it as a path
            try:
                os.makedirs(value)
            except OSError as e:
                if e.errno != errno.EEXIST:
                    raise

        # sync both environment variables and houdini vars
        # because houdini doesn't do sync os value when opening new files
        # we force it to the current value if current value already matches
        # the new value on the hscript var
        current_value = hou.hscript("echo -n `${}`".format(var))[0]
        os.environ[var] = current_value

        if current_value != value:
            hou.hscript("set {}={}".format(var, value))
            os.environ[var] = value

            print("  - Updated ${} to {}".format(var, value))
```

Note the **TODO** blocks that I added which might still be relevant for you. also, I haven't used this before so please do check that if we expose this whether e.g. setting var `$ASSET_FRAMESTART` to `{assetData[frameStart]}` actually makes it so that you can use it as an int where we want to? If not, we might even need to set the correct type, etc. Anyway, have fun puzzling! :)